src/client/clientmap.cpp

```diff
@@ -288,206 +283,289 @@ void ClientMap::updateDrawList()
 
 	auto is_frustum_culled = m_client->getCamera()->getFrustumCuller();
 
-	// Uncomment to debug occluded blocks in the wireframe mode
-	// TODO: Include this as a flag for an extended debugging setting
-	// if (occlusion_culling_enabled && m_control.show_wireframe)
-	// 	occlusion_culling_enabled = porting::getTimeS() & 1;
+	std::set<v3s16> shortlist;
+	MeshGrid mesh_grid = m_client->getMeshGrid();
 
-	std::queue<v3s16> blocks_to_consider;
+	/*
+		When range_all is enabled, simply enumerate all blocks visible in the
+		frustum and display them.
+	 */
+	if(m_control.range_all) {
+		for (auto &sector_it : m_sectors) {
+			MapSector *sector = sector_it.second;
+			if (!sector)
+				continue;
 
-	// Bits per block:
-	// [ visited | 0 | 0 | 0 | 0 | Z visible | Y visible | X visible ]
-	MapBlockFlags blocks_seen(p_blocks_min, p_blocks_max);
+			MapBlockVect sectorblocks;
+			sector->getBlocks(sectorblocks);
 
-	// Start breadth-first search with the block the camera is in
-	blocks_to_consider.push(camera_block);
-	blocks_seen.getChunk(camera_block).getBits(camera_block) = 0x07; // mark all sides as visible
+			/*
+				Loop through blocks in sector
+			*/
+			for (MapBlock *block : sectorblocks) {
+				MapBlockMesh *mesh = block ? block->mesh : nullptr;
 
-	std::set<v3s16> shortlist;
-	MeshGrid mesh_grid = m_client->getMeshGrid();
+				// Calculate the coordinates for range and frutum culling
+				v3f mesh_sphere_center;
+				f32 mesh_sphere_radius;
 
-	// Recursively walk the space and pick mapblocks for drawing
-	while (blocks_to_consider.size() > 0) {
+				v3s16 block_pos_nodes = block->getPos() * MAP_BLOCKSIZE;
 
-		v3s16 block_coord = blocks_to_consider.front();
-		blocks_to_consider.pop();
+				if (mesh) {
+					mesh_sphere_center = intToFloat(block_pos_nodes, BS)
+							+ mesh->getBoundingSphereCenter();
+					mesh_sphere_radius = mesh->getBoundingRadius();
+				}
+				else {
+					mesh_sphere_center = intToFloat(block_pos_nodes, BS) + v3f((MAP_BLOCKSIZE * 0.5f - 0.5f) * BS);
+					mesh_sphere_radius = 0.0f;
+				}
+				// Frustum culling
+				// Only do coarse culling here, to account for fast camera movement.
+				// This is needed because this function is not called every frame.
+				float frustum_cull_extra_radius = 300.0f;
+				if (is_frustum_culled(mesh_sphere_center,
+						mesh_sphere_radius + frustum_cull_extra_radius))
+					continue;
```

Comment:
can we avoid duplicating this?

Reply:
I do not see how.

Reply:
In fact seeing how the old culler is actually faster when there are many empty blocks not sent from the server, I want to bring that one back anyway.

Reply:
See also #13169